File: client/src/screen/AchievementsPanel.java
package screen;

import java.awt.Component;
import java.util.ArrayList;

import javax.swing.JPanel;

import object.AchievementBadge;

public class AchievementsPanel extends JPanel
{
	String panelName = "";
	
	public AchievementsPanel(String panelName)
	{
		super();
		
		this.panelName = panelName;
	}
	
	public ArrayList<AchievementBadge> getAchievementBadges()
	{
		ArrayList<AchievementBadge> ret = new ArrayList<>();
		
		Component[] components = getComponents();
		for (int i=0; i<components.length; i++)
		{
			Component c = components[i];
			if (c instanceof AchievementBadge)
			{
				AchievementBadge badge = (AchievementBadge)c;
				ret.add(badge);
			}
		}
		
		return ret;
	}
	
	public String getPageTitle()
	{
		ArrayList<AchievementBadge> badges = getAchievementBadges();
		int totalForPage = badges.size();
		
		int totalUnlocked = 0;
		for (int i=0; i<badges.size(); i++)
		{
			AchievementBadge badge = badges.get(i);
			if (badge.isEarned())
			{
				totalUnlocked++;
			}
		}
		
		return panelName + " (" + totalUnlocked + "/" + totalForPage + ")";
	}
}

File: client/src/main/kotlin/screen/AchievementsPanel.kt
package screen

import bean.AchievementBadge
import javax.swing.JPanel
import utils.Achievement

const val MAX_ACHIEVEMENTS_PER_PAGE = 36

fun makeAchievementPanels() =
    Achievement.entries.chunked(MAX_ACHIEVEMENTS_PER_PAGE).mapIndexed { ix, badges ->
        AchievementsPanel(ix + 1, badges.map(::AchievementBadge))
    }

class AchievementsPanel(private val number: Int, private val badges: List<AchievementBadge>) :
    JPanel() {
    init {
        badges.forEach(::add)
    }

    fun getTitle(): String {
        val total = badges.count { it.isVisible }
        return "Page $number (${badges.count { it.isEarned }} / $total)"
    }
}
